### PR TITLE
Gen 4 Pokémon Mystery Gift Overhaul

### PR DESCRIPTION
**Pull request by ubergeek77**
I've gone in and "overhauled" the way Gen 4 Mystery Gifts are distributed to connected clients. Here's an overview of what I've changed:

- 19 out of the full list of 40 Gen 4 Pokémon Game IDs were previously missing entirely. These games have been added to the code and are now recognized as Gen 4 Pokémon games for Mystery Gift support.
  - For instance, before these changes, Pokémon Pearl and Pokémon HeartGold were not recognized as Gen 4 Pokémon games.
- Rather than distribute the gifts based on a date moduli, every Gen 4 Mystery Gift is distributed based on its original, real life Nintendo Wi-Fi Connection distribution date, on a per Game ID basis.
  - Each game language will receive the appropriate Wonder Card in that language.
  - Since the Nintendo DS' clock is used for the "received" date on a Wonder Card, this distribution method makes it easier for players to receive Mystery Gifts that are indistinguishable from gifts that were actually downloaded from the real Nintendo Wi-Fi Connection.
- Only events that were specifically *Wi-Fi Events* have been included in this overhaul.
  - Because there are ***substantially*** more *Local Events* that have been distributed in the past, because said events are largely disorganized between different languages and locales, and because the distribution times for those would have overlapped with the ones I have included in the code, I did not include any *Local Events* in this overhaul. This is an emulator for Nintendo Wi-Fi Connection, after all.
- This repository had some .myg files that did not correspond to real-life Nintendo Wi-Fi Connection distributions. As to not delete them outright and preserve them, they have been moved to a separate subfolder.
  - If there is ever a point where it is desired to keep the old distribution method with these old .myg files, it would be trivial to add a special case for, for instance, Nintendo DS clock dates where the year is greater than or equal to 2019. Connections with local dates matching that year could simply fall back to the day modulus method, where the variable `dlc_path` would be appended before calling the relevant functions. Doing this does make the code *slightly* less easy to read, so while I have tested this myself outside of my repository, I did not find it appropriate to include in this pull request.
 - The new .myg files that I have added use filenames that are much easier to understand. They also specify the language for that file, where previous filenames found in this repository did not. For example:
   - Previous Secret Key filename: *17p.myg*
   - My Secret Key File: *ENG_017_Plat_SecretKey.myg*
- I'm not entirely sure if GitHub will include my Wiki in this pull request, but I have also updated the relevant Wiki pages for this change. These pages will tell users how the new distribution system works, as well as provide a full, comprehensive list of the events that will be distributed. I have changed one page and added another page:
  - Changed: [Nintendo DS Download Content](https://github.com/ubergeek77/dwc_network_server_emulator/wiki/Nintendo-DS-Download-Content)
  - Added: [Generation 4 Pokémon Mystery Gifts](https://github.com/ubergeek77/dwc_network_server_emulator/wiki/Generation-4-Pok%C3%A9mon-Mystery-Gifts)

I have tested this new method on my own devices, and it works great. I am very happy with these changes.

I appreciate your consideration into this pull request. If there is anything unsatisfactory about my code changes, please let me know, and I will do my best to address your concerns.

### CODE REVIEW

**Review by sepalani — CHANGES_REQUESTED** [review]
This PR can be a bit improved, IMO.

It doesn't seem PEP8 compliant and I'm also not fond of these if-else-if forests, a (named) tuple, a dictionary or another data structure can be used instead and you can loop through it.  I'll think about it and add some details regarding that later this week.

Then, The first and last commit can be squashed together.

Otherwise, I'm not against the idea. If you need some help on how to squash these commits or use git in general, feel free to ask.

**Comment by ubergeek77** [comment]
Very fair points. This can absolutely be improved to be more dynamic. It would be much better to have some sort of "event registry" that has all of the needed information on an event, and then have the code generate the needed Game IDs and filenames on the fly, rather than being manually specified. Perhaps all of this could even be read from a .json file.

I have a few ideas on how to tackle this, so I'll take a crack at it later today.

**Comment by ubergeek77** [comment]
I've moved the event definitions out of the code and into a json file. Here's how it works:

- The client DS requests a Mystery Gift from the server.
- The server loops over every event found in the .json file.
- For any given event definition:
  - If the current Game ID's *language subcode* (last character in Game ID) is found in the *lang_subcodes* key for said event, continue to the next check.
  - If the current Game ID matches the "short code" contained in the *valid_games* key for said event (where a short code is a 1 character representation of a Gen 4 game), continue to the next check.
  - The server retrieves the given event's registered start date and end date, and determines if the Nintendo DS' reported date is within those two dates. If it is, we have found our event. We continue to filename generation:
  - Using the following items in the following order, dynamically generate the filename we expect for said event:
    - Current Game ID's *language subcode* converted to a 3-character identifier (i.e. J -> JPN)
    - Current event's *Wonder Card ID* (padded to 3 characters)
    - A human-readable representation of the list of games the event is valid for
    - The event's registered *shortname*
  - The logic used in the previous point to generate the expected filename will match every .myg file I previously submitted in the PR.

The benefit of moving this to a .json file is that anyone can add their own events rather easily, so long as the .myg filename matches what the server expects for that event. Care will have to be taken to ensure the start and end dates for each event do not overlap with any others, and I have done so for the events in my submitted `g4_events.json` file; no event will prevent another event from being served.

Additionally, the file `dlc.py` has been formatted for PEP8 compliance.

Please let me know what you think!

**Comment by ubergeek77** [comment]
Quick update: I wrote a verification script using the new function I made, and there are some slight errors in my .json file. I also caught a few .myg files that are missing. I'll go through this and commit the fixes soon.

**Comment by ubergeek77** [comment]
Aforementioned event errors have been corrected.

**Comment by barronwaffles** [comment]
I'll free up some time over the next few days to look over the changes, but the idea of using the original distribution dates does feel like a practical solution.

File: other/dlc.py [hunk 1]
```diff
@@ -21,35 +21,63 @@
...
-gamecodes_return_random_file = [
-    'ADAD',
+# Gen 4 Pokemon Games (for Mystery Gift Distribution)
+# If a game from this list requests a file listing, the server will compare
+# the client Nintendo DS's date setting to the list of Mystery Gifts
+# that were actually distributed in real life at the given time.
+# This way, if someone knows exactly when an event happened in real life,
+# they know exactly which date to change to in order to receive
+# a particular gift.
+gen_4_pokemon_gamecodes = [
```
Comment (barronwaffles) [inline other/dlc.py:40]: This list could probably be collapsed further by excluding the language key/indicator - it doesn't seem like it's being utilized at this level.

File: other/dlc.py [hunk 2]
```diff
@@ -117,6 +145,82 @@ def filter_list_by_date(data, token):
     return ret
 
 
+def todays_g4_event_filename(today, gamecd, dlc_path):
+    event_filename = "null"
```
Comment (barronwaffles) [inline other/dlc.py:149]: It might save some logic if you returned 'None' when the event evaluation below failed

**Review by barronwaffles — COMMENTED** [review]
I'll just throw a couple of comments now - still too busy to scan over it much further, sorry.

Will attempt to free up this weekend for further review.

**Comment by ubergeek77** [comment]
Hello friends,

I know it's been a while, and as it happens I am now quite busy with my own work, but I will gladly address the minor comments @barronwaffles made, as well as anything else he finds, upon his further review.

I'm actually going to be away on a business trip for the next few weeks (woo...), but I'm hoping work will calm down after that. If I have more things to fix by that time, you can hopefully expect a revised PR from me.

**Comment by barronwaffles** [comment]
Sorry for the delay on my end - struggling to find any free time myself.

**Comment by barronwaffles** [comment]
Pinging due to DMCA shenanigans - will finally check this out soon.

**Comment by fancythedeveloper** [comment]
Sucks that AltWFC for like a month got removed because of a false DMCA claim. FFS

**Pull request closed without merging**